mm2src/crypto/src/global_hd_ctx.rs

```diff
@@ -0,0 +1,105 @@
+use crate::privkey::{bip39_priv_key_from_seed, key_pair_from_secret, PrivKeyError};
+use crate::{mm2_internal_der_path, Bip32DerPathOps, Bip32Error, Bip44PathToCoin, CryptoInitError, CryptoInitResult};
+use bip32::{ChildNumber, ExtendedPrivateKey};
+use keys::{KeyPair, Private, Public as PublicKey, Secret as Secp256k1Secret};
+use mm2_err_handle::prelude::*;
+use std::convert::TryInto;
+use std::num::TryFromIntError;
+use std::ops::Deref;
+use std::sync::Arc;
+
+const HARDENED: bool = true;
+const NON_HARDENED: bool = false;
+
+#[derive(Clone)]
+pub struct GlobalHDAccountArc(Arc<GlobalHDAccountCtx>);
+
+impl Deref for GlobalHDAccountArc {
+    type Target = GlobalHDAccountCtx;
+
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+pub struct GlobalHDAccountCtx {
+    bip39_priv_key: ExtendedPrivateKey<secp256k1::SecretKey>,
+    /// secp256k1 key pair derived at `mm2_internal_der_path`.
+    /// It's used for mm2 internal purposes such as signing P2P messages.
+    mm2_internal_key_pair: KeyPair,
+    /// `hd_account` actually means an `address` index at the `m/purpose'/coin'/account'/chain/address` path.
+    /// This account is set globally for every activated coin.
+    hd_account: ChildNumber,
+}
+
+impl GlobalHDAccountCtx {
+    pub fn new(passphrase: &str, hd_account_id: u64) -> CryptoInitResult<GlobalHDAccountCtx> {
+        let bip39_priv_key = bip39_priv_key_from_seed(passphrase)?;
+
+        let hd_account_id =
+            hd_account_id
+                .try_into()
+                .map_to_mm(|e: TryFromIntError| CryptoInitError::InvalidHdAccount {
+                    hd_account_id,
+                    error: e.to_string(),
+                })?;
+        let hd_account =
+            ChildNumber::new(hd_account_id, NON_HARDENED).map_to_mm(|e| CryptoInitError::InvalidHdAccount {
```

Comment:
Is there a reason you chose to use `NON_HARDENED` here since it's less secure according to this https://bitcoin.stackexchange.com/a/37489
There is two quotes from the linked discussion that I would like to mention here:
1 - "non-hardened public keys are weaker when an attacker has 1) the extended public key and 2) one of the non-hardened private keys that was derived from it. In that circumstance, the attacker can work out the private key of the extended public key, and therefore get every key that can be derived from that, hardened and non-hardened."
2 - "Hardened keys are useful whenever you don't need the ability to give someone a master public key, and have them know all addresses. For those cases, they have much better security properties than unhardened."

Reply:
Good question!
We need to provide backwards compatible with other wallets.
`Global HD account` feature should allow the user to use `BIP39` mnemonic passphrase, and should allow to derive addresses exactly the same scheme as other wallets.
[BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#path-levels) determines the following derivation path:
> m / purpose' / coin_type' / account' / change / address_index

In our case, coins have `m/purpose'/coin_type'` in their config where `purpose` and `coin_type` are always hardened.
Then to activate a coin, we use `account = 0` (always hardened), `change = 0` (non-hardened) and `address_index = ctx.conf.hd_account_id` (non-hardened).

Yes, in this scheme if a user revealed his private key of an activated coin, for example, `m/44'/141'/0'/0/1` where `ctx.conf.hd_account_id = 1`. Then an attacker can work out an extended private keys of the nearest hardened child (it's `account` in the derivation path). So then the attacker can derive private keys for every `chain` and `account_index` children.

@tonymorony @artemii235 we would reduce the attack surface if we mark all children as hardened in the derivation path `m / purpose' / coin_type' / account' / change' / address_index'`, but then mm2 wouldn't be compatible with other wallets.
What do you think?

Reply:
Thanks for the explanation :)
My opinion on this is that compatibility with other wallets is more important, I think users would never reveal a private key anyways since it's equivalent to revealing the mnemonic passphrase (Metamask and other wallets give warnings when a user wants to export an account private key).

Reply:
Our main goal of this integration was to be compatible with other wallets. I guess it is actual still, so we have to use what others use.